### PR TITLE
Updates for FE POST orchestration

### PR DESCRIPTION
**Pull request by BenCosfordHMRC**
https://jira.tools.tax.service.gov.uk/browse/SASS-2602

Added start of post orchestration - added validation on the period data so that:
Theres no duplicates, not more than 12 months, not an empty submission etc. 

##### Before Reviewing
- [x]  Have you pulled the branch down?
- [x]  Have you assigned yourself to the PR?
- [x]  Have you moved the task to “in review” on JIRA?
- [ ]  Have you checked to ensure all dependencies are up to date?
- [x]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [x]  Have you run the tests?
- [x]  Have you run the journey tests?
- [x]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [x]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [x]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [x]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [x]  Have you run the tests?
- [x]  Have you run the journey tests? (where applicable)
- [x]  Have you addressed warnings where appropriate?
- [x]  Have you rebased against the current version of main?
- [x]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [x]  Have you checked the PR Builder passes?

### CODE REVIEW

File: test/services/ContractorCYAServiceSpec.scala [hunk 1]
```diff
@@ -16,33 +16,62 @@
 
 package services
 
+import models.{APIErrorBodyModel, APIErrorModel, HttpParserError, InvalidOrUnfinishedSubmission}
 import models.mongo.DataNotUpdatedError
+import play.api.http.Status.INTERNAL_SERVER_ERROR
+import support.builders.models.CISSubmissionBuilder.{aCreateCISSubmission, aPeriodData, anUpdateCISSubmission}
```
Comment (ina-tu) [inline test/services/ContractorCYAServiceSpec.scala:22]: unused import

File: test/services/DeleteCISPeriodServiceSpec.scala [hunk 2]
```diff
@@ -16,18 +16,86 @@
 
 package services
 
+import models.{APIErrorBodyModel, APIErrorModel, HttpParserError, InvalidOrUnfinishedSubmission}
+import play.api.http.Status.INTERNAL_SERVER_ERROR
+import support.builders.models.AllCISDeductionsBuilder.anAllCISDeductions
+import support.builders.models.CISSourceBuilder.aCISSource
+import support.builders.models.CISSubmissionBuilder.{aPeriodData, anUpdateCISSubmission}
```
Comment (ina-tu) [inline test/services/DeleteCISPeriodServiceSpec.scala:23]: unused import

File: test/support/mocks/MockCISSessionService.scala [hunk 3]
```diff
@@ -17,10 +17,9 @@
 package support.mocks
 
 import java.time.Month
-
 import models.mongo.{CisCYAModel, CisUserData, DatabaseError}
 import models.{HttpParserError, IncomeTaxUserData, ServiceError, User}
-import org.scalamock.handlers.{CallHandler3, CallHandler4, CallHandler6}
+import org.scalamock.handlers.{CallHandler3, CallHandler4, CallHandler5, CallHandler6}
```
Comment (ina-tu) [inline test/support/mocks/MockCISSessionService.scala:22]: unused import

File: test/support/mocks/MockContractorCYAService.scala [hunk 4]
```diff
@@ -16,10 +16,12 @@
 
 package support.mocks
 
+import models.mongo.CisUserData
 import models.{ServiceError, User}
-import org.scalamock.handlers.CallHandler3
+import org.scalamock.handlers.{CallHandler3, CallHandler5}
```
Comment (ina-tu) [inline test/support/mocks/MockContractorCYAService.scala:21]: unused import

File: test/support/mocks/MockDeleteCISPeriodService.scala [hunk 5]
```diff
@@ -16,10 +16,11 @@
 
 package support.mocks
 
-import models.{ServiceError, User}
-import org.scalamock.handlers.CallHandler4
+import models.{IncomeTaxUserData, ServiceError, User}
+import org.scalamock.handlers.{CallHandler4, CallHandler6}
```
Comment (ina-tu) [inline test/support/mocks/MockDeleteCISPeriodService.scala:20]: unused import

File: test/support/mocks/MockDeleteCISPeriodService.scala [hunk 6]
```diff
@@ -32,9 +33,10 @@ trait MockDeleteCISPeriodService extends MockFactory {
                              employerRef: String,
                              user: User,
                              deductionPeriod: Month,
-                             result: Either[ServiceError, Unit]): CallHandler4[Int, String, User, Month, Future[Either[ServiceError, Unit]]] = {
-    (mockService.removeCisDeduction(_: Int, _: String, _: User, _: Month))
-      .expects(taxYear, employerRef, user, deductionPeriod)
+                             result: Either[ServiceError, Unit]): CallHandler6[Int, String, User,
```
Comment (ina-tu) [inline test/support/mocks/MockDeleteCISPeriodService.scala:36]: should IncomeTaxUserData be passed as a parameter here?

File: test/utils/DateMonthUtilSpec.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package utils
+
+import org.scalatestplus.play.guice.GuiceOneAppPerSuite
```
Comment (ina-tu) [inline test/utils/DateMonthUtilSpec.scala:19]: unused import

**Review by ina-tu — CHANGES_REQUESTED** [review]

File: test/controllers/DeleteCISPeriodControllerSpec.scala [hunk 8]
```diff
@@ -51,15 +52,23 @@ class DeleteCISPeriodControllerSpec extends ControllerUnitTest
...
       mockRemoveCISDeduction(taxYearEOY, aCisDeductions.employerRef, aUser, MAY, Right(()))
 
       status(underTest.submit(taxYearEOY, aCisDeductions.employerRef, month = "may").apply(fakeIndividualRequest)) shouldBe SEE_OTHER
     }
 
     "throw an error when service returns Left" in {
-      mockUserPriorDataForEOY(taxYearEOY, aCisDeductions.employerRef, month = "may", anIncomeTaxUserData.copy(cis = Some(anAllCISDeductions)))
+      mockExclusivelyCustomerPriorDataForEOY(taxYearEOY, aCisDeductions.employerRef, month = "may", anIncomeTaxUserData.copy(cis = Some(anAllCISDeductions)))
       mockRemoveCISDeduction(taxYearEOY, aCisDeductions.employerRef, aUser, MAY, Left(HttpParserError(INTERNAL_SERVER_ERROR)))
+      mockHandleError(INTERNAL_SERVER_ERROR, InternalServerError)
```
[the remark this answers is not in the document]
Comment (BenCosfordHMRC) [reply test/controllers/DeleteCISPeriodControllerSpec.scala:64]: this case is defaulting to internal error page if mongo fails or other error

File: test/support/mocks/MockDeleteCISPeriodService.scala (hunk 6 above)
Comment (BenCosfordHMRC) [reply test/support/mocks/MockDeleteCISPeriodService.scala:36]: for the controller tests, not that interested in what the data is, so left it out.

**Review by ina-tu — APPROVED** [review]

**Review by ina-tu — APPROVED** [review]

**Pull request merged**